URL globbing

roniemartinez opened the pull request:
Implements `curl` globbing: https://everything.curl.dev/cmdline/globbing

- [x] List
- [x] Numerical ranges
- [x] Alphabetical ranges
- [x] Combinations
- [x] Zero-prefixes for numerical ranges

Resolves https://github.com/spider-rs/spider/issues/111

j-mendez commented:
@roniemartinez good work, if you need a review for anything feel free to ping me :). Thanks for taking this on!

roniemartinez commented:
@j-mendez I believe all desired behavior are here with tests.

But not sure whether we change the types (or how and why) to e.g. CompactString or u32 to another type. I am just starting Rust and I am doing very basic things. Please take a look at this draft.

j-mendez commented:
@roniemartinez pushed the changes to compact string. We can now dump the list into the crawler for the initial run.

j-mendez commented:
@roniemartinez I can add the initial expansion into the `website.rs` now if you want.

roniemartinez commented:
@j-mendez yes please, I can see it's on the "links" block but still figuring it out. Please start.

j-mendez commented:
> @j-mendez yes please, I can see it's on the "links" block but still figuring it out. Please start.

Will have something pushed soon, after dinner.

j-mendez commented:
@roniemartinez changes pushed `cargo test --verbose --features glob`. All tests pass. Feel free to make any adjustments before sending to review. Great work on feature and clean code + test!

j-mendez commented:
@roniemartinez I moved itertools into scope incase we refactor outside and just use the cfg macro in this one location.

j-mendez commented:
Looks, like some things broke. On it now.

roniemartinez commented:
@j-mendez I added a few tests.

Overall I think, I am good to put this for review.

j-mendez commented:
@roniemartinez ok, we just need to adjust one last thing to get the pr in.`expand_url` needs to decode urls since CompactString adds  `%7D/` ex: `"http://choosealicense.com/licenses/%7Bmit,apache-2.0,mpl-2.0%7D/"`. Can fix this shortly.

roniemartinez commented:
@j-mendez wouldn't `CompactString` fail the scenario when I use `%7B` and `%7D`?

https://everything.curl.dev/cmdline/globbing#using-in-urls

e.g. According to the link if I want to use `{` and `}` without globbing, `expand_url` should not treat it as a list: `http://choosealicense.com/licenses/%7Bmit%7D/` should remain as is.

j-mendez commented:
> http://choosealicense.com/licenses/%7Bmit,apache-2.0,mpl-2.0%7D/

For some reason it was not returning correctly, we can add a test to help too.

roniemartinez commented:
@j-mendez I added the test. `decode` is breaking it so I removed it in https://github.com/spider-rs/spider/pull/113/commits/f860a4b113c44ac580adaa05a6b52435ac087ebf

roniemartinez commented on .github/workflows/bench.yml:
```diff
@@ -9,6 +9,10 @@ env:
   CARGO_TERM_COLOR: always
   RUST_LOG: "off"
 
+concurrency:
+  group: ${{ github.ref }}
+  cancel-in-progress: true
+
```
I added this since the action takes too long

roniemartinez replied:
The workflow needs your approval @j-mendez

j-mendez commented:
> @j-mendez I added the test. `decode` is breaking it so I removed it in [f860a4b](https://github.com/spider-rs/spider/commit/f860a4b113c44ac580adaa05a6b52435ac087ebf)

Ok, we need to update the logic in one case if it is empty so the test pass here ` merino --no-auth & ./target/debug/spider_worker & cargo test --verbose --all-features`

j-mendez commented:
@roniemartinez the issue is we cant use the url like that to perform the network request. We need to decode after the expansion.

j-mendez commented:
> @roniemartinez the issue is we cant use the url like that to perform the network request. We need to decode after the expansion.

Link: "http://choosealicense.com/licenses/%7Bmit,apache-2.0,mpl-2.0%7D/" - page Page { html: None, links: {} }

test:

start `spider_worker` local first and `merino --no-auth` proxy

` cargo test --verbose --all-features website::test_crawl_glob -- --nocapture`

roniemartinez commented:
@j-mendez I think the issue is that `expand_url` is receiving an encoded URL which it shouldn't be. It should receive the raw URL.

```
received: http://choosealicense.com/licenses/%7Bmit,apache-2.0,mpl-2.0%7D/
```

j-mendez commented:
> @j-mendez I think the issue is that `expand_url` is receiving an encoded URL which it shouldn't be. It should receive the raw URL.
> 
> ```
> received: http://choosealicense.com/licenses/%7Bmit,apache-2.0,mpl-2.0%7D/
> ```

So this test case should return a value by default when not encoded?

```rust
#[cfg(feature = "glob")]
#[test]
fn test_expand_url_percent_encoded() {
    let url = "https://choosealicense.com/licenses/%7Bmit%7D/";

    assert_eq!(expand_url(url), Vec::<CompactString>::new());
}

```

roniemartinez commented:
> So this test case should return a value by default when not encoded?

Yes, it is related to this: https://everything.curl.dev/cmdline/globbing#using-in-urls

j-mendez commented:
> > @j-mendez I think the issue is that `expand_url` is receiving an encoded URL which it shouldn't be. It should receive the raw URL.
> > ```
> > received: http://choosealicense.com/licenses/%7Bmit,apache-2.0,mpl-2.0%7D/
> > ```
> 
> So this test case should return a value by default when not encoded?
> 
> ```rust
> #[cfg(feature = "glob")]
> #[test]
> fn test_expand_url_percent_encoded() {
>     let url = "https://choosealicense.com/licenses/%7Bmit%7D/";
> 
>     assert_eq!(expand_url(url), Vec::<CompactString>::new());
> }
> ```

I put the fallback when it is empty to be the url passed. I realized that it could be an invalid url depending on if it was a proper blob. A better fallback would be to use the base domain when all else fails.

roniemartinez commented:
Encoding seems to happen in `match url::Url::parse(&self.domain) {}` block.

roniemartinez commented:
@j-mendez Tests passes in 692775ff5e6c25a4769861d1a6371c1d4293b808

j-mendez commented:
> @j-mendez Tests passes in [692775f](https://github.com/spider-rs/spider/commit/692775ff5e6c25a4769861d1a6371c1d4293b808)

This [commit](https://github.com/spider-rs/spider/pull/113/commits/e0fa788cb0fb321b69e411bda2996f43f020751a) fixes the test based on if someone puts an incorrect or empty glob list to fallback to root domain. The crawl needs to succeed at all cost ;P

j-mendez commented:
> @j-mendez Tests passes in [692775f](https://github.com/spider-rs/spider/commit/692775ff5e6c25a4769861d1a6371c1d4293b808)

Great change and good catch finding this. The parity on the crawls should be the same now with this too, thank you!

roniemartinez commented:
@j-mendez I think there is a missing import on the last commit

j-mendez commented:
> @j-mendez I think there is a missing import on the last commit

ahh, on it now thanks.

j-mendez commented:
@roniemartinez amazing work on everything done. If I do not merge this in tonight, ill take a look in the AM. Sorry for rushing a bit on some of the parts hope it was not too painful on getting the feature in!

roniemartinez commented:
I am still getting error when running the test locally, should we make `convert_abs_path` public?

```
error[E0425]: cannot find function `convert_abs_path` in this scope
   --> spider/src/website.rs:491:35
    |
491 |                     expanded.push(convert_abs_path(&u, "/").as_str().into());
    |                                   ^^^^^^^^^^^^^^^^ not found in this scope
    |
note: function `crate::page::convert_abs_path` exists but is inaccessible
   --> spider/src/page.rs:67:1
    |
67  | fn convert_abs_path(base: &Url, href: &str) -> Url {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible

For more information about this error, try `rustc --explain E0425`.
error: could not compile `spider` due to previous error
```

j-mendez commented:
> I am still getting error when running the test locally, should we make `convert_abs_path` public?
> 
> ```
> error[E0425]: cannot find function `convert_abs_path` in this scope
>    --> spider/src/website.rs:491:35
>     |
> 491 |                     expanded.push(convert_abs_path(&u, "/").as_str().into());
>     |                                   ^^^^^^^^^^^^^^^^ not found in this scope
>     |
> note: function `crate::page::convert_abs_path` exists but is inaccessible
>    --> spider/src/page.rs:67:1
>     |
> 67  | fn convert_abs_path(base: &Url, href: &str) -> Url {
>     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible
> 
> For more information about this error, try `rustc --explain E0425`.
> error: could not compile `spider` due to previous error
> ```
> 
> EDIT: Fixed in [ae60e67](https://github.com/spider-rs/spider/commit/ae60e67337e21b34d1aa804228cdc400efdc5c61)

Yes, thank you. Rushed the gun again 😅.

roniemartinez commented:
> If I do not merge this in tonight, ill take a look in the AM. Sorry for rushing a bit on some of the parts hope it was not too painful on getting the feature in!

No worries, take your time. Thank you for helping implement this feature 🙇
